### PR TITLE
EIP-1186: add `eth_getProof` RPC-Method

### PR DESCRIPTION
**Pull request by simon-jentzsch**
Implementation for the [EIP-1186](https://github.com/ethereum/EIPs/issues/1186)

Ethereum uses MerkleTrees to store the state of accounts and their storage. This allows verification of each value by simply creating a MerkleProof. But currently the eth-Module does not give you access to these proofs. This EIP suggests a additional RPC-Method, which creates MerkleProofs for Accounts and Storage-Values. 

Combined with a stateRoot (from the blockheader) it enables offline verification of any account or storage-value. This allows especially IOT-Devices or even mobile apps which are not able to run a light client to verify responses from a untrusted source.

### CODE REVIEW

**Comment by 5chdn** [comment]
Hey, please sign our CLA.

> This EIP suggests an additional RPC-Method, which creates MerkleProofs for Accounts and Storage-Values.

Which EIP are you referring to? Is this officially part of the eth namespace?

File: docker/hub/Dockerfile [hunk 1]
```diff
@@ -35,10 +35,9 @@ RUN apt-get update && \
 # show backtraces
  RUST_BACKTRACE=1 && \
 # build parity
-cd /build&&git clone https://github.com/paritytech/parity && \
+cd /build&&git clone https://github.com/slockit/parity -b in3 && \
```
Comment (5chdn) [inline docker/hub/Dockerfile]: Please undo the changes to the Dockerfile, otherwise you break our release scripts.

File: ethcore/src/state/account.rs [hunk 2]
```diff
@@ -477,7 +477,7 @@ impl Account {
 		use trie::recorder::Recorder;
 
 		let mut recorder = Recorder::new();
-
+
```
Comment (5chdn) [inline ethcore/src/state/account.rs]: Please remove trailing whitespaces.

**Comment by simon-jentzsch** [comment]
Thanks for the responses. 
This referrs to the https://github.com/ethereum/EIPs/issues/1186 we just started. 
I'll check the CLA und of course I'll also fix the changes we made for our docker-script.

**Comment by 5chdn** [comment]
Ah, gotcha. I'll mark this WIP.

File: rpc/src/v1/impls/eth.rs [hunk 3]
```diff
@@ -26,7 +26,11 @@ use parking_lot::Mutex;
 
 use ethash::SeedHashCompute;
 use ethcore::account_provider::{AccountProvider, DappId};
-use ethcore::client::{BlockChainClient, BlockId, TransactionId, UncleId, StateOrBlock, StateClient, StateInfo, Call, EngineInfo};
+use ethcore::client::{BlockChainClient, BlockId, TransactionId, UncleId, StateOrBlock, StateClient, StateInfo, Call, EngineInfo, ProvingBlockChainClient};
+
+//use ethcore::client::{BlockChainClient, BlockId, TransactionId, UncleId, StateOrBlock, StateClient, StateInfo, Call, EngineInfo, ProvingBlockChainClient};
```
Comment (dvdplm) [inline rpc/src/v1/impls/eth.rs]: remove

File: rpc/src/v1/impls/eth.rs [hunk 4]
```diff
@@ -557,6 +562,54 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+			Some(p) => Ok(EthAccount {
+				 address : _address.into(),
+				 balance : p.1.balance.into(),
+				 nonce : p.1.nonce.into(),
+				 code_hash : p.1.code_hash.into(),
+				 storage_hash : p.1.storage_root.into(),
+				 account_proof: Some(p.0.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()), 
+				 storage_proof: Some(
+					 values.iter().map(|storageIndex| { 
+						 let key2 : H256 = storageIndex.clone().into();
```
Comment (dvdplm) [inline rpc/src/v1/impls/eth.rs]: whitespace

File: rpc/src/v1/impls/light/eth.rs [hunk 5]
```diff
@@ -493,6 +493,12 @@ impl<T: LightChainClient + 'static> Eth for EthClient<T> {
 		}))
 	}
 
+
+	fn proof(&self, address: RpcH160, values:Vec<RpcH256>, num: Trailing<BlockNumber>) -> BoxFuture<EthAccount> {
+		Box::new(future::err(errors::unimplemented(None)))
```
Comment (dvdplm) [inline rpc/src/v1/impls/light/eth.rs]: Intentional?

File: rpc/src/v1/types/account_info.rs [hunk 6]
```diff
@@ -13,13 +13,44 @@
...
+use v1::types::{ H160, H256, U256, Bytes};
 
 /// Account information.
 #[derive(Debug, Default, Clone, PartialEq, Serialize)]
 pub struct AccountInfo {
 	/// Account name
 	pub name: String,
 }
+#[derive(Debug, Default, Clone, PartialEq, Serialize)]
+pub struct StorageProof{
```
Comment (dvdplm) [inline rpc/src/v1/types/account_info.rs]: docs for this?

**Comment by dvdplm** [comment]
The CI failure is spurious here and should be fixed if you pull in the latest changes from master.

File: rpc/src/v1/impls/light/eth.rs (hunk 5 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/light/eth.rs]: At the moment getting a proof within the light-client would require a full state, but I agree we would need to implement this in the light_fetcher to retrieve the proof from a peer with a full state. This still needs to be done.

**Comment by simon-jentzsch** [comment]
[clabot:check]

**Comment by simon-jentzsch** [comment]
Thanks, Yes, I also considered using a different namespace. In the end I believe that offering the proof as a rpc-function should be part of the standard-module. That's why I choose the `eth_` namespace, on the other hand it would be easier to have this function in one of the next releases, if this is a part of the `parity` -module. But in the long run I want all clients to support this. But I'm open for discussion. And until we get this feature in, we are using the forked-version.

**Comment by 5chdn** [comment]
Ok, let me know how you want to proceed.

**Comment by sorpaas** [comment]
@simon-jentzsch We can use the `parity_` namespace first and then move it to `eth_` namespace once it's standardized. This is how `eth_chainId`/`parity_chainId` was dealt with.

**Comment by simon-jentzsch** [comment]
@sorpaas Yes, this would be an option to use the parity-namespace, but since it is also important that geth  will implement it, I rather have it as a real standard. For now we can live with using our fork of parity. I am also maintaining Docker-Images https://hub.docker.com/r/slockit/parity-in3/tags/ which we are using. But I just hope the EIP will go through soon. ;-)

**Comment by simon-jentzsch** [comment]
@5chdn I created the PR for the EIP and hope that it will go through. I also merged the changes for parity 2.x and created a Docker-Container on dockerhub. So I think it would be good,if we can repopen this PR.

**Comment by simon-jentzsch** [comment]
@5chdn Would be good to reopen this pull request, since the EIP is now merged and a offical Draft: https://eips.ethereum.org/EIPS/eip-1186

**Comment by 5chdn** [comment]
ok :+1: 

please rebase on master

File: rpc/src/v1/types/account_info.rs (hunk 6 above)
Comment (simon-jentzsch) [reply rpc/src/v1/types/account_info.rs]: done

**Comment by simon-jentzsch** [comment]
thanks, So I just rebased the branch. Looks good.

File: rpc/src/v1/impls/eth.rs [hunk 7]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
 		Box::new(future::done(res))
 	}
 
+	fn proof(&self, _address: RpcH160, values:Vec<RpcH256>, num: Trailing<BlockNumber>) -> BoxFuture<EthAccount> {
```
Comment (sorpaas) [inline rpc/src/v1/impls/eth.rs]: nit: missing a space in `values:` and `Vec`.

File: rpc/src/v1/impls/eth.rs [hunk 8]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+		let a : H160 = _address.clone().into();
```
Comment (sorpaas) [inline rpc/src/v1/impls/eth.rs]: We usually have space after `:` and type, but not before.

File: rpc/src/v1/impls/eth.rs (hunk 7 above)
Comment (sorpaas) [inline rpc/src/v1/impls/eth.rs]: `_address` is used, so it shouldn't start with `_`.

File: rpc/src/v1/impls/eth.rs [hunk 9]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+		let res = match self.client.prove_account(key1, id) {
+			Some(p) => Ok(EthAccount {
+				 address : _address.into(),
+				 balance : p.1.balance.into(),
+				 nonce : p.1.nonce.into(),
+				 code_hash : p.1.code_hash.into(),
+				 storage_hash : p.1.storage_root.into(),
+				 account_proof: Some(p.0.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
+				 storage_proof: Some(
+					 values.iter().map(|storage_index| {
```
Comment (sorpaas) [inline rpc/src/v1/impls/eth.rs]: The below lines' indentations are really weird.

File: rpc/src/v1/impls/light/eth.rs [hunk 10]
```diff
@@ -47,7 +47,7 @@ use v1::helpers::light_fetch::{self, LightFetch};
 use v1::traits::Eth;
 use v1::types::{
 	RichBlock, Block, BlockTransactions, BlockNumber, LightBlockNumber, Bytes, SyncStatus, SyncInfo,
-	Transaction, CallRequest, Index, Filter, Log, Receipt, Work,
+	Transaction, CallRequest, Index, Filter, Log, Receipt, Work,EthAccount,
```
Comment (sorpaas) [inline rpc/src/v1/impls/light/eth.rs]: Missing space

File: rpc/src/v1/impls/eth.rs (hunk 7 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 8 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 7 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 9 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/light/eth.rs (hunk 10 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/light/eth.rs]: fixed

File: rpc/src/v1/types/account_info.rs [hunk 11]
```diff
@@ -13,13 +13,44 @@
 
 // You should have received a copy of the GNU General Public License
 // along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+use v1::types::{ H160, H256, U256, Bytes};
```
Comment (niklasad1) [inline rpc/src/v1/types/account_info.rs]: nit extra space

File: rpc/src/v1/impls/eth.rs [hunk 12]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+
+		try_bf!(check_known(&*self.client, num.clone()));
+		let res = match self.client.prove_account(key1, id) {
+			Some(p) => Ok(EthAccount {
+				 address : address.into(),
+				 balance : p.1.balance.into(),
+				 nonce : p.1.nonce.into(),
+				 code_hash : p.1.code_hash.into(),
+				 storage_hash : p.1.storage_root.into(),
+				 account_proof: Some( p.0.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
```
[the remark this answers is not in the document]
Comment (niklasad1) [reply rpc/src/v1/impls/eth.rs]:
```suggestion
				 account_proof: Some(p.0),
```

File: rpc/src/v1/impls/eth.rs [hunk 13]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+			BlockNumber::Pending => {
+				warn!("`Pending` is deprecated and may be removed in future versions. Falling back to `Latest`");
+				BlockId::Latest
+			}
+		};
+
+		try_bf!(check_known(&*self.client, num.clone()));
+		let res = match self.client.prove_account(key1, id) {
+			Some(p) => Ok(EthAccount {
+				 address : address.into(),
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: needless `into()`

File: rpc/src/v1/impls/eth.rs [hunk 14]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+				 balance : p.1.balance.into(),
+				 nonce : p.1.nonce.into(),
+				 code_hash : p.1.code_hash.into(),
+				 storage_hash : p.1.storage_root.into(),
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: Needless `into()` for all

File: rpc/src/v1/types/account_info.rs [hunk 15]
```diff
@@ -13,13 +13,44 @@
...
 
 /// Account information.
 #[derive(Debug, Default, Clone, PartialEq, Serialize)]
 pub struct AccountInfo {
 	/// Account name
 	pub name: String,
 }
+/// Datastructure with proof for one single storage-entry
```
Comment (niklasad1) [inline rpc/src/v1/types/account_info.rs]: empty line between the end-parenthesis and doc comment, please

File: rpc/src/v1/traits/eth.rs [hunk 16]
```diff
@@ -63,6 +63,10 @@ build_rpc_trait! {
 		#[rpc(name = "eth_getBalance")]
 		fn balance(&self, H160, Trailing<BlockNumber>) -> BoxFuture<U256>;
 
+		/// Returns balance of the given account.
```
Comment (niklasad1) [inline rpc/src/v1/traits/eth.rs]: This description is not accurate because it actually does much more than that. i.e, includes Merkle-proofs for both the account and the storage.

Why as you stated in the EIP: `Returns the account- and storage-values of the specified account including the Merkle-proof`?

File: rpc/src/v1/impls/eth.rs [hunk 17]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+				 account_proof: Some( p.0.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
+				 storage_proof: Some( values.iter().map(|storage_index| {
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: Would be better to use `filter_map` here and only append valid proofs IMHO

**Comment by niklasad1** [comment]
@simon-jentzsch 

Are you working on this? 

Let us know if you need help with anything or if something is not clear thanks!

File: rpc/src/v1/types/account_info.rs [hunk 18]
```diff
@@ -13,13 +13,44 @@
...
+	pub balance: U256,
+	#[serde(rename="nonce")]
+	pub nonce: U256,
+	#[serde(rename="codeHash")]
+	pub code_hash: H256,
+	#[serde(rename="storageHash")]
+	pub storage_hash: H256,
+  #[serde(rename="accountProof")]
+	pub account_proof: Option<Vec<Bytes>>,
+  #[serde(rename="storageProof")]
```
[the remark this answers is not in the document]
Comment (simon-jentzsch) [reply rpc/src/v1/types/account_info.rs]: fixed

File: rpc/src/v1/impls/eth.rs [hunk 19]
```diff
@@ -545,6 +546,55 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+			Some(p) => Ok(EthAccount {
+				 address : address.into(),
+				 balance : p.1.balance.into(),
+				 nonce : p.1.nonce.into(),
+				 code_hash : p.1.code_hash.into(),
+				 storage_hash : p.1.storage_root.into(),
+				 account_proof: Some( p.0.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
+				 storage_proof: Some( values.iter().map(|storage_index| { 
+						 let key2 : H256 = storage_index.clone().into();
+					   match self.client.prove_storage(key1, keccak(key2), id) {
```
[the remark this answers is not in the document]
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed all spaces to tabs

File: rpc/src/v1/impls/eth.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: Well not exactly, I actually get a `Vec<Vec<u8>>` which I need to convert to `Vec<Bytes>`.
If you can suggest a more efficient way to convert or cast, let me know.

File: scripts/docker/alpine/Dockerfile [hunk 20]
```diff
@@ -33,7 +33,7 @@ RUN addgroup -g 1000 parity \
 
 USER parity
 
-EXPOSE 8080 8545 8180
+EXPOSE 5001 8080 8082 8083 8545 8546 8180 30303/tcp 30303/udp
```
[the remark this answers is not in the document]
Comment (simon-jentzsch) [reply scripts/docker/alpine/Dockerfile]: ok, reverted the ports.

File: rpc/src/v1/impls/eth.rs (hunk 14 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: The reason I use `into()` is because the I need to convert `ethereum_types::H256` into `v1::types::hash::H256`

Without the `into()` I would get a compile-error.

File: rpc/src/v1/types/account_info.rs (hunk 15 above)
Comment (simon-jentzsch) [reply rpc/src/v1/types/account_info.rs]: fixed

File: rpc/src/v1/traits/eth.rs (hunk 16 above)
Comment (simon-jentzsch) [reply rpc/src/v1/traits/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 17 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: changed it, because even for a not existing `0x0` - storage value  there must a proof, so if we don't get a storage proof we skip these keys. (which actually should never happen, since this is only possible if the account does not exist in which case I currently get at least an empty proof)

File: rpc/src/v1/types/account_info.rs (hunk 11 above)
Comment (simon-jentzsch) [reply rpc/src/v1/types/account_info.rs]: fixed space

File: rpc/src/v1/impls/eth.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (niklasad1) [reply rpc/src/v1/impls/eth.rs]: Sorry I got confused with the `Bytes type` in ```prove_account```  but it was only a type-alias
```suggestion
   account_proof: Some(proof.into_iter().map(Bytes::new).collect(),
```
Instead of cloning the bytes move them instead.

File: rpc/src/v1/impls/eth.rs (hunk 14 above)
Comment (niklasad1) [reply rpc/src/v1/impls/eth.rs]: Sorry, right, but you can remove ```into()``` on the address

File: rpc/src/v1/impls/eth.rs [hunk 21]
```diff
@@ -452,7 +453,7 @@ fn check_known<C>(client: &C, number: BlockNumber) -> Result<()> where C: BlockC
 const MAX_QUEUE_SIZE_TO_MINE_ON: usize = 4;	// because uncles go back 6.
 
 impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<C, SN, S, M, EM> where
-	C: miner::BlockChainClient + BlockChainClient + StateClient<State=T> + Call<State=T> + EngineInfo + 'static,
+  C: miner::BlockChainClient + StateClient<State=T> + ProvingBlockChainClient + Call<State=T> + EngineInfo + 'static,
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: tabs please

File: rpc/src/v1/impls/eth.rs [hunk 22]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+			BlockNumber::Pending => {
+				warn!("`Pending` is deprecated and may be removed in future versions. Falling back to `Latest`");
+				BlockId::Latest
+			}
+		};
+
+		try_bf!(check_known(&*self.client, num.clone()));
+		let res = match self.client.prove_account(key1, id) {
+			Some((proof,account)) => Ok(EthAccount {
+				address : address.into(),
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]:
```suggestion
				address,
```

File: rpc/src/v1/impls/eth.rs [hunk 23]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+				balance : account.balance.into(),
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: formatting,  should be (same for all other fields)

```suggestion
				balance: account.balance.into(),
```

File: rpc/src/v1/impls/eth.rs [hunk 24]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+				nonce : account.nonce.into(),
+				code_hash : account.code_hash.into(),
+				storage_hash : account.storage_root.into(),
+				account_proof: Some( proof.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]:
```suggestion
				account_proof: Some(proof.into_iter().map(Bytes::new).collect()),
```

File: rpc/src/v1/impls/eth.rs [hunk 25]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+				storage_proof: Some( values.iter().filter_map(|storage_index| {
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]: take ownership and skip cloning

```suggestion
				storage_proof: Some(values.into_iter().filter_map(|storage_index| { 
```

File: rpc/src/v1/impls/eth.rs [hunk 26]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+		let res = match self.client.prove_account(key1, id) {
+			Some((proof,account)) => Ok(EthAccount {
+				address : address.into(),
+				balance : account.balance.into(),
+				nonce : account.nonce.into(),
+				code_hash : account.code_hash.into(),
+				storage_hash : account.storage_root.into(),
+				account_proof: Some( proof.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()),
+				storage_proof: Some( values.iter().filter_map(|storage_index| { 
+					let key2 : H256 = storage_index.clone().into();
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]:
```suggestion
					let key2: H256 = storage_index.into();
```

File: rpc/src/v1/impls/eth.rs [hunk 27]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+					match self.client.prove_storage(key1, keccak(key2), id) {
+							Some((storage_proof,storage_value)) =>  Some(StorageProof {
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]:
```suggestion
							Some((storage_proof,storage_value)) => Some(StorageProof {
```

File: rpc/src/v1/impls/eth.rs [hunk 28]
```diff
@@ -545,6 +546,51 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+								key : key2.into(),
+								value: storage_value.into(),
+								proof: storage_proof.iter().map(|b| Bytes::new(b.clone())).collect::<Vec<Bytes>>()
```
Comment (niklasad1) [inline rpc/src/v1/impls/eth.rs]:
```suggestion
																proof: storage_proof.into_iter().map(Bytes::new).collect(),
```

File: rpc/src/v1/impls/light/eth.rs [hunk 29]
```diff
@@ -470,6 +470,12 @@ impl<T: LightChainClient + 'static> Eth for EthClient<T> {
 		}))
 	}
 
+
+	fn proof(&self, _address: RpcH160, _values:Vec<RpcH256>, _num: Trailing<BlockNumber>) -> BoxFuture<EthAccount> {
+		Box::new(future::err(errors::unimplemented(None)))
+	}
+
+
```
Comment (niklasad1) [inline rpc/src/v1/impls/light/eth.rs]: remove duplicate empty line

File: rpc/src/v1/types/account_info.rs [hunk 30]
```diff
@@ -21,6 +22,28 @@ pub struct AccountInfo {
...
+#[serde(rename_all = "camelCase")]
+pub struct EthAccount {
+	pub address: H160,
+	pub balance: U256,
+	pub nonce: U256,
+	pub code_hash: H256,
+	pub storage_hash: H256,
+	pub account_proof: Option<Vec<Bytes>>,
+	pub storage_proof: Option<Vec<StorageProof>>,
+}
```
Comment (niklasad1) [inline rpc/src/v1/types/account_info.rs]: Doesn't make sense to me that `account_proof` is an Option because `EthAccount` is only constructed if `prove_account` returns `Some(proof, account)`?

File: rpc/src/v1/impls/eth.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: right, that works! Changed it.

File: rpc/src/v1/impls/eth.rs (hunk 21 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 22 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: right, fixed

File: rpc/src/v1/impls/eth.rs (hunk 23 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 24 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 25 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 26 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 27 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/eth.rs (hunk 28 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: fixed

File: rpc/src/v1/impls/light/eth.rs (hunk 29 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/light/eth.rs]: removed

File: rpc/src/v1/types/account_info.rs (hunk 30 above)
Comment (simon-jentzsch) [reply rpc/src/v1/types/account_info.rs]: this is true, so I removed the option. Also for `storage_proof` since this will return an empty array if no keys are specified.

File: rpc/src/v1/impls/eth.rs (hunk 14 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: yes, fixed the address

File: rpc/src/v1/impls/eth.rs (hunk 26 above)
Comment (niklasad1) [reply rpc/src/v1/impls/eth.rs]: Change `values.into_iter().filter_map(..)` and it should work

Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: filter_map is already used for `storage_proof`, but the storageIndex can't be converted without the clone :

``` 
// does not work since storage_index is  a &v1::types::hash::H256, but I need a ethereum_types::H256
let key2: H256 = storage_index.into();
```

Comment (niklasad1) [reply rpc/src/v1/impls/eth.rs]: hrm, maybe I don't understand but this compiles for me:

```rustc
values 
        .into_iter()
        .filter_map(|storage_index| {
            let key2: H256 = storage_index.into();
            match self.client.prove_storage(key1, keccak(key2), id) {
                Some((storage_proof, storage_value)) => Some(StorageProof {
                    key: key2.into(),
                    value: storage_value.into(),
                    proof: storage_proof.into_iter().map(Bytes::new).collect(),
                }),
                None => None,
            }
        })
        .collect::<Vec<StorageProof>>(),
```

Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: You are right, the `into_iter` was the point I was missing. Now it also compiles on my computer.
Yes, I'm still tryng to get used to rust, but I'm getting therer. Thanks!

**Comment by Tbaut** [comment]
Could you please update the doc in src/interfaces/eth.js interfaces https://github.com/parity-js/jsonrpc

**Comment by simon-jentzsch** [comment]
@niklasad1  I fixed the formatting of the structs and also added tests.
@Tbaut Added `eth_getProof` at the doc. ( See PR https://github.com/parity-js/jsonrpc/pull/14 )

**Comment by simon-jentzsch** [comment]
Thanks, I'm looking forward to see this merged!

**Comment by tomusdrw** [comment]
IMHO this should be enabled only with `--jsonrpc-experimental`, see #9928

**Comment by simon-jentzsch** [comment]
- tests are running again
- implemented the check for the experimental RPCs flag

File: rpc/src/v1/impls/eth.rs [hunk 31]
```diff
@@ -547,6 +551,52 @@ impl<C, SN: ?Sized, S: ?Sized, M, EM, T: StateInfo + 'static> Eth for EthClient<
...
+			Some((proof,account)) => Ok(EthAccount {
+				address: address,
+				balance: account.balance.into(),
+				nonce: account.nonce.into(),
+				code_hash: account.code_hash.into(),
+				storage_hash: account.storage_root.into(),
+				account_proof: proof.into_iter().map(Bytes::new).collect(),
+				storage_proof: values.into_iter().filter_map(|storage_index| { 
+					let key2: H256 = storage_index.into();
+					match self.client.prove_storage(key1, keccak(key2), id) {
```
Comment (tomusdrw) [inline rpc/src/v1/impls/eth.rs]: could be a `.map` instead of `match`

File: rpc/src/v1/tests/eth.rs [hunk 32]
```diff
@@ -198,6 +204,33 @@ fn eth_get_balance() {
 	assert_eq!(tester.handler.handle_request_sync(req_new_acc).unwrap(), res_new_acc);
 }
 
+#[test]
+fn eth_get_proof() {
```
Comment (tomusdrw) [inline rpc/src/v1/tests/eth.rs]: Would be good to have a test already when `experimental_rpcs` are disabled.

**Comment by simon-jentzsch** [comment]
Yes, getting a lot of proof could be heavy, but since you can't get the entire storage without specifying all available storage keys, it is only heavy, if you pass a long list of keys. 
So it may make sense for infura to limit the number of storage values, but if you running your own node and want to get all cryptokitties, then you should know what you are doing. ( and how to calculate all storage keys )

https://youtu.be/26ucTSSaqog?t=7448

File: rpc/src/v1/impls/eth.rs (hunk 31 above)
Comment (simon-jentzsch) [reply rpc/src/v1/impls/eth.rs]: ok, changed it.

**Comment by simon-jentzsch** [comment]
But concerning the performance, I think there is room for optimization here (which can be done later ).

after getting the account-proof we already have the BasicAccount-Object and so the `storage_root`.

But when calling the `prove_storage` (just like the lightclient), this function always fetches the Account-Object again before finding the storage key. And this happens for each storage key.
So it would be more efficient to get Account only once.

https://github.com/paritytech/parity-ethereum/blob/master/ethcore/src/state/mod.rs#L1203
